src/components/membership-queue.ts

```diff
@@ -279,11 +287,19 @@ export class MembershipQueue {
         }
     }
 
-    private shouldRetry(ex: {body: {code: string; errcode: string;}, statusCode: number}, attempts: number): boolean {
-        return !(
-            attempts === this.opts.maxAttempts ||
-            ex.body.errcode === "M_FORBIDDEN" ||
-            ex.statusCode === 403
-        );
+    private shouldRetry(ex: Error|MatrixError|{statusCode: number}, attempts: number): boolean {
+        if (attempts >= this.opts.maxAttempts) {
+            // Do not retry if we're out of attempts
+            return false;
+        }
+        if (ex instanceof MatrixError) {
+            // Only retry if the call wasn' forbidden
+            return ex.errcode !== "M_FORBIDDEN";
```

Comment:
I'd expect us to check for `M_LIMIT_EXCEEDED` here – any reason to retry others?